url90b6689baa_adsonrocha_PD321_tgz-pJ8-MyFuzzyLiteClassJ8/src/com/fuzzylite/Console.java

```diff
@@ -0,0 +1,623 @@
+                engine = mamdani();
+            } else if (example.equals("t") || example.equals("ts")
+                    || example.equals("takagi-sugeno")) {
+                engine = takagiSugeno();
+            } else {
+                throw new RuntimeException(String.format(
+                        "[option error] example <%s> not available", example));
+            }
+            inputFormat = "fll";
+            textEngine.append(new FllExporter().toString(engine));
+        } else {
+
+            String inputFilename = options.get(KW_INPUT_FILE);
+            if (inputFilename == null) {
+                throw new RuntimeException("[option error] no input file specified");
+            }
+            File inputFile = new File(inputFilename);
+            if (!inputFile.exists()) {
+                inputFile.createNewFile();
+            }
+
+            BufferedReader reader = new BufferedReader(new FileReader(inputFile));
+            try {
+                String line = reader.readLine();
+                while (line != null) {
+                    textEngine.append(line).append("\n");
+                    line = reader.readLine();
+                }
+            } catch (Exception ex) {
+                throw new RuntimeException(ex);
+            } finally {
+                reader.close();
+            }
+
+            inputFormat = options.get(KW_INPUT_FORMAT);
+            if (inputFormat == null || inputFormat.isEmpty()) {
+                int extensionIndex = inputFilename.lastIndexOf(".");
+                if (extensionIndex >= 0) {
+                    inputFormat = inputFilename.substring(extensionIndex + 1);
+                } else {
+                    throw new RuntimeException("[format error] unspecified format of input file");
+                }
+            }
+        }
+
+        String outputFilename = options.get(KW_OUTPUT_FILE);
+        String outputFormat = options.get(KW_OUTPUT_FORMAT);
+        if (outputFormat == null || outputFormat.isEmpty()) {
+            if (outputFilename == null || outputFilename.isEmpty()) {
+                throw new RuntimeException("[format error] unspecified format of output");
+            } else {
+                int extensionIndex = outputFilename.lastIndexOf(".");
+                if (extensionIndex >= 0) {
+                    outputFormat = outputFilename.substring(extensionIndex + 1);
+                } else {
+                    throw new RuntimeException("[format error] unspecified format of output file");
+                }
+            }
+        }
+
+        Writer writer;
+        if (outputFilename == null || outputFilename.isEmpty()) {
+            writer = new StringWriter();
+        } else {
+            File outputFile = new File(outputFilename);
+            if (!outputFile.exists()) {
+                outputFile.createNewFile();
+            }
+            writer = new FileWriter(outputFile);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *writer*. The resource is referenced at lines 235, 236, 241, 243. The resource is closed at lines 238, 246. There are other execution paths that do not contain closure statements, for example, when Writer.flush throws an exception, and an else condition at line 270 evaluates to true. Ensure *writer* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).